core/arch/arm/plat-stm32mp1/plat_huk.c

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ * Copyright (c) 2022, Foundries.io Limited
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <crypto/crypto.h>
+#include <drivers/stm32_bsec.h>
+#include <kernel/tee_common_otp.h>
+#include <platform_config.h>
+#include <string.h>
+
+#if !defined(CFG_STM32MP1_HUK_BSEC_BASE)
+#error "CFG_STM32MP1_HUK_BSEC_BASE undefined"
+#endif
+
+static TEE_Result aes_gcm_encrypt_uid(uint8_t *key, size_t key_len,
+				      uint8_t *out, size_t *out_len)
+{
+	uint8_t uid[HW_UNIQUE_KEY_LENGTH] = { 0 };
+	uint8_t nonce[12] = {
+		0xd2, 0x45, 0x0e, 0x07, 0xea, 0x5d,
+		0xe0, 0x42, 0x6c, 0x0f, 0xa1, 0x33,
+	};
+	uint8_t tag[16] = { 0 };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t tag_len = sizeof(tag);
+	size_t uid_len = sizeof(uid);
+	uint32_t word = 0;
+	void *ctx = NULL;
+	size_t n = 0;
+
+	for (n = 0; n < (UID2_OTP - UID0_OTP + 1); n++) {
+		res = stm32_bsec_read_otp(&word, UID0_OTP + n);
+		if (res)
+			return res;
+
+		memcpy(uid + n * sizeof(uint32_t), &word, sizeof(uint32_t));
+	}
+
+	/* HUK is the AES-GCM encrypted UID using a secret key from OTP */
+	res = crypto_authenc_alloc_ctx(&ctx, TEE_ALG_AES_GCM);
+	if (res)
+		return res;
+
+	res = crypto_authenc_init(ctx, TEE_MODE_ENCRYPT, key, key_len,
+				  nonce, sizeof(nonce),
+				  TEE_AES_BLOCK_SIZE, 0, uid_len);
+	if (res)
+		return res;
+
+	res = crypto_authenc_enc_final(ctx, uid, sizeof(uid), out, out_len,
+				       tag, &tag_len);
+	if (res)
+		return res;
+
+	crypto_authenc_final(ctx);
+	crypto_authenc_free_ctx(ctx);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	uint32_t bsec_base = CFG_STM32MP1_HUK_BSEC_BASE;
+	uint8_t otp_data[HW_UNIQUE_KEY_LENGTH] = { 0 };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t len = HW_UNIQUE_KEY_LENGTH;
+	uint32_t word = 0;
+	uint32_t tmp = 0;
+	bool lock = false;
+	size_t n = 0;
+	static bool initialized;
+	const char *state = "Locked";
+	const char *type = "OTP";
+
+	/* Read 128 bits of OTP data (the secret) */
+	for (n = 0; n < HW_UNIQUE_KEY_LENGTH / sizeof(uint32_t); n++) {
+		res = stm32_bsec_read_permanent_lock(bsec_base + n, &lock);
+		if (res) {
+			EMSG("huk: read permanent lock");
+			return res;
+		}
+
+		if (!lock)
+			state = "Devel";
+
+		res = stm32_bsec_read_otp(&tmp, bsec_base + n);
+		if (res)
+			return res;
+
+		res = stm32_bsec_shadow_register(bsec_base + n);
+		if (res) {
+			EMSG("huk: access shadow register");
+			return res;
+		}
+
+		res = stm32_bsec_read_otp(&word, bsec_base + n);
+		if (res) {
+			EMSG("huk: read OTP secret");
+			return res;
+		}
+
+		res = stm32_bsec_write_otp(tmp, bsec_base + n);
```

Comment:
Why this process of loading (read shadow mem then reload shadow mem and write shadow into shadow mem?

Could call `stm32_bsec_shadow_otp()` once at at initialization (to load shadow mem with fuse data) and use `stm32_bsec_read_otp()` for each call to read the shadowed data.

Reply:
Not exactly. https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/plat-stm32mp1/nsec-service/bsec_svc.c#L41-L50 attempts the read the fuse, bypassing shadow mem.

Mainline OP-TEE/stm32mp1 has no specific strategy on using the BSEC locks. OP-TEE components only read fuse values, that is load fuse in shadow mem and read them from that shadow memory. The values you can read from fuses depend on the state of the several BSEC locks. If an early boot stage or an earlier power cycle sequence burned some BSEC locks, OP-TEE will end in using the values found in shadow mem that may differ from fuses effective values. There are 4 locks around OTPs (with few specific constraints). One is persistent across power cylces and forbids programming of a fuse unit (permanent_lock). 3 locks are sticky (unlocked at system reset): one to forbid raw programming of shadow memory (sw_lock), one to forbid reloading of fuse in shadow mem (sr_lock) and last one to forbid programming of fuse unit (sp_lock, it's a sticky variant of permanent lock).

So if you want to ensure using fuses definitive values:
- check permanent_lock is set for each unit read.
- check sw_lock is set for each unit read.
- load fuse units in shardow mem and read from there.
- lock sw_lock if you want to forbid further modification on HUK value
- if all permanent_lock's and sw_lock's were set and we've locked sw_lock's
  then we're using "Locked" keys, otherwise, we're using "Devel" keys.